src/core/elevation/qgsprofilerenderer.cpp

```diff
@@ -0,0 +1,194 @@
+#include "qgsprofilerenderer.h"
+#include "qgsabstractprofilesource.h"
+#include "qgsabstractprofilegenerator.h"
+#include "qgscurve.h"
+
+#include <QtConcurrentMap>
+#include <QtConcurrentRun>
+
+QgsProfilePlotRenderer::QgsProfilePlotRenderer( const QList< QgsAbstractProfileSource * > &sources,
+    const QgsProfileRequest &request )
+  : mRequest( request )
+{
+  for ( QgsAbstractProfileSource *source : sources )
+  {
+    if ( source )
+    {
+      if ( std::unique_ptr< QgsAbstractProfileGenerator > generator{ source->createProfileGenerator( mRequest ) } )
+        mGenerators.emplace_back( std::move( generator ) );
+    }
+  }
+}
+
+QgsProfilePlotRenderer::~QgsProfilePlotRenderer()
+{
+  if ( isActive() )
+  {
+    cancelGeneration();
+  }
+}
+
+void QgsProfilePlotRenderer::startGeneration()
+{
+  if ( isActive() )
+    return;
+
+  mStatus = Generating;
+
+  Q_ASSERT( mJobs.empty() );
+
+  mJobs.reserve( mGenerators.size() );
+  for ( const auto &it : mGenerators )
+  {
+    ProfileJob job;
+    job.generator = it.get();
+    mJobs.emplace_back( std::move( job ) );
+  }
+
+  connect( &mFutureWatcher, &QFutureWatcher<void>::finished, this, &QgsProfilePlotRenderer::onGeneratingFinished );
+
+  mFuture = QtConcurrent::map( mJobs, generateProfileStatic );
+  mFutureWatcher.setFuture( mFuture );
+}
+
+void QgsProfilePlotRenderer::cancelGeneration()
+{
+  if ( !isActive() )
+    return;
+
+  disconnect( &mFutureWatcher, &QFutureWatcher<void>::finished, this, &QgsProfilePlotRenderer::onGeneratingFinished );
+
+  for ( const ProfileJob &job : mJobs )
+  {
+    if ( job.generator )
+    {
+      if ( QgsFeedback *feedback = job.generator->feedback() )
+      {
+        feedback->cancel();
+      }
+    }
+  }
+
+  mFutureWatcher.waitForFinished();
+
+  onGeneratingFinished();
+}
+
+void QgsProfilePlotRenderer::cancelGenerationWithoutBlocking()
+{
+  if ( !isActive() )
+    return;
+
+  for ( const ProfileJob &job : mJobs )
+  {
+    if ( job.generator )
+    {
+      if ( QgsFeedback *feedback = job.generator->feedback() )
+      {
+        feedback->cancel();
+      }
+    }
+  }
+}
+
+void QgsProfilePlotRenderer::waitForFinished()
+{
+  if ( !isActive() )
+    return;
+
+  disconnect( &mFutureWatcher, &QFutureWatcher<void>::finished, this, &QgsProfilePlotRenderer::onGeneratingFinished );
+  mFutureWatcher.waitForFinished();
+
+  onGeneratingFinished();
+}
+
+bool QgsProfilePlotRenderer::isActive() const
+{
+  return mStatus != Idle;
+}
+
+QgsDoubleRange QgsProfilePlotRenderer::zRange() const
+{
+  double min = std::numeric_limits< double >::max();
+  double max = std::numeric_limits< double >::lowest();
+  for ( const ProfileJob &job : mJobs )
+  {
+    if ( job.complete && job.results )
+    {
+      const QgsDoubleRange jobRange = job.results->zRange();
+      min = std::min( min, jobRange.lower() );
+      max = std::max( max, jobRange.upper() );
+    }
+  }
+  return QgsDoubleRange( min, max );
+}
```

Comment:
If there are no jobs or no results (raster layer not set as a elevation surface or profile layers turned off) then this can return 
```
QgsDoubleRange( std::numeric_limits< double >::max(), std::numeric_limits< double >::lowest() )
```
(`QgsDoubleRange` seems to have no problem with that) giving inf/-inf values to the plot's mMinY/mMaxY, eventually  leading to endless loops on zoom full.

Some rational default return values when `min > max` could solve that.

Reply:
Ah, **great** catch! I'll fix.

Reply:
Fixed